### DISCUSSION TITLE
How to use React context API in remix?

### DISCUSSION THREAD

**Discussion by tiranuom** [remix-run/remix#1785] [Q&A] [answered]
Hi All,

I am new to remix and was playing with it. I needed to use React context API within remix application and when I implement, server side rendering part works perfectly. however in the client side rendering part it hangs the application throwing errors continuously.

I need to know whether,

1. React context API is supported in remix?
2. If supported is there any example I can refer?

Thanks in advance.

**Comment by sergiodxa** [comment, marked as answer]
It should work as usual

1. Create a context object with React.createContext (do it outside route files)
2. In some part render the context.Provider with some value
3. In another part call useContext(context) to read the value

Also note that the Outlet component has a context prop (`<Outlet context={value} />`) and a pair hook useOutletContext which let you share values between a parent and a child route.

And, also note that if you want to access another route (loader) data you can use the useMatches hook to get all the routes data, and you can use [useRouteData from Remix Utils](https://github.com/sergiodxa/remix-utils#useroutedata) which uses useMatches to get the data too.

**Reply by tiranuom** [reply]
Thank you very much @sergiodxa, After your comment, I got it working. 

And I found my issue also. I have implemented the context creation logic in a `*.server.ts` file, which was not available in the client code. Silly me.

**Reply by felixmpaulus** [reply]
Using the react api I get hydration errors when rendering the context's values in an outlet. 
Using the context prop on the outlet works though.

**Reply by BluntKatana** [reply]
Hello @sergiodxa, we (at our startup) were wondering why it is that a React.createContext needs to be created outside a route file? Could you maybe explain this behaviour perhaps? :)

**Reply by sergiodxa** [reply]
This is a little bit old, and it may have changed with Vite, but the issue was that since routes are entry points in the bundle, the code there is not moved to a separate chunk after build. This means the file importing the context has a copy of the context object, and the file creating the context and rendering the provider has another copy, since there are two copies there are two different context objects and they can't share the value.

**Comment by sayhicoelho** [comment]
I'm receiving an error `ERROR: Expected ">" but found "value"`  when trying to import my ThemeProvider using React Context API.

![image](https://github.com/remix-run/remix/assets/8188365/ce12ad4f-c92a-49e2-a35a-e80aea7cf5be)

![image](https://github.com/remix-run/remix/assets/8188365/d1030c3f-a4b8-4a4f-9b12-6dcf7c9f7e69)

**Reply by Fantros** [reply]
I have the same problem, did you solve it?

**Reply by sayhicoelho** [reply]
You can't import .js files. You must import .ts files. The Remix's documentation is messy.

**Reply by lanesky** [reply]
I found a simpler workaround  - changing ".js" to ".jsx" for example `theme.js` to `theme.jsx`, that will skip the eslint.
